pkg/containers-publisher/containers-publisher.go

```diff
@@ -0,0 +1,135 @@
+	log "github.com/sirupsen/logrus"
+
+	containercollection "github.com/inspektor-gadget/inspektor-gadget/pkg/container-collection"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+)
+
+type ContainersPublisher struct {
+	collection *containercollection.ContainerCollection
+
+	containersDs              datasource.DataSource
+	eventTypeField            datasource.FieldAccessor
+	cgroupIDField             datasource.FieldAccessor
+	mountNsIDField            datasource.FieldAccessor
+	nameField                 datasource.FieldAccessor
+	containersSubscriptionKey string
+}
+
+func NewContainersPublisher(gadgetCtx operators.GadgetContext, collection *containercollection.ContainerCollection) (*ContainersPublisher, error) {
+	publisher := &ContainersPublisher{}
+	var err error
+
+	publisher.containersDs, err = gadgetCtx.RegisterDataSource(datasource.TypeSingle, "containers")
+	if err != nil {
+		return nil, fmt.Errorf("creating datasource: %w", err)
+	}
+	publisher.containersDs.AddAnnotation("cli.default-output-mode", "none")
+
+	publisher.eventTypeField, err = publisher.containersDs.AddField("event_type", api.Kind_String)
+	if err != nil {
+		return nil, fmt.Errorf("adding field event_type: %w", err)
+	}
+
+	publisher.cgroupIDField, err = publisher.containersDs.AddField("cgroup_id", api.Kind_Uint64)
+	if err != nil {
+		return nil, fmt.Errorf("adding field cgroup_id: %w", err)
+	}
+
+	publisher.mountNsIDField, err = publisher.containersDs.AddField("mntns_id", api.Kind_Uint64)
+	if err != nil {
+		return nil, fmt.Errorf("adding field mntns_id: %w", err)
+	}
+
+	publisher.nameField, err = publisher.containersDs.AddField("name", api.Kind_String)
+	if err != nil {
+		return nil, fmt.Errorf("adding field name: %w", err)
+	}
+
+	publisher.collection = collection
+
+	return publisher, nil
+}
+
+func (publisher *ContainersPublisher) emitContainersDatasourceEvent(eventType containercollection.EventType, container *containercollection.Container, k8s bool) error {
+	ev, err := publisher.containersDs.NewPacketSingle()
+	if err != nil {
+		return fmt.Errorf("creating new containers datasource packet: %w", err)
+	}
+
+	publisher.eventTypeField.PutString(ev, eventType.String())
+	publisher.cgroupIDField.PutUint64(ev, container.CgroupID)
+	publisher.mountNsIDField.PutUint64(ev, container.Mntns)
+	if k8s {
+		publisher.nameField.PutString(ev, container.K8s.ContainerName)
+	} else {
+		publisher.nameField.PutString(ev, container.Runtime.ContainerName)
+	}
+
+	err = publisher.containersDs.EmitAndRelease(ev)
+	if err != nil {
+		return fmt.Errorf("emitting containers datasource event: %w", err)
+	}
+
+	return nil
+}
+
+func (publisher *ContainersPublisher) PublishContainers(k8s bool, host bool, fakeContainer *containercollection.Container, containerSelector containercollection.ContainerSelector) error {
```

Comment:
nit: Add Start() to indicate it actually starts something that happens in the future (emit container events as they appear).

```suggestion
func (publisher *ContainersPublisher) Start(k8s bool, host bool, fakeContainer *containercollection.Container, containerSelector containercollection.ContainerSelector) error {
```

Reply:
Same for unsusbscribe, what about calling it Stop()?

Reply:
Having this as an operator would be great, but I fear we still have some limitations to make it possible, like allowing a parameter to be shared among multiple operators (the filter parameters in this case).